lenskit-core/src/main/java/org/lenskit/data/dao/file/TextEntitySource.java

```diff
@@ -0,0 +1,284 @@
+ */
+public class TextEntitySource implements EntitySource {
+    private static final Logger logger = LoggerFactory.getLogger(TextEntitySource.class);
+    private final String name;
+    private CharSource source;
+    private URL sourceURL;
+    private EntityFormat format;
+
+    /**
+     * Construct a new text entity source.
+     */
+    public TextEntitySource() {
+        this("<unnamed>");
+    }
+
+    /**
+     * Construct a new text entity source.
+     * @param name The source's name.
+     */
+    public TextEntitySource(String name) {
+        this.name = name;
+    }
+
+    /**
+     * Get the name of this data source.
+     * @return The data source name.
+     */
+    public String getName() {
+        return name;
+    }
+
+    /**
+     * Set the source file for this reader.
+     * @param file The source file.
+     */
+    public void setFile(Path file) {
+        source = Files.asCharSource(file.toFile(), Charset.defaultCharset());
+        try {
+            sourceURL = file.toUri().toURL();
+        } catch (MalformedURLException e) {
+            throw new IllegalArgumentException("invalid URL " + file, e);
+        }
+    }
+
+    /**
+     * Set the URL of the input data.
+     * @param url The URL of the input data.
+     */
+    public void setURI(URL url) {
+        sourceURL = url;
+        source = Resources.asCharSource(url, Charsets.UTF_8);
+    }
+
+    /**
+     * Get the source file for this data source.
+     * @return The source file.
+     */
+    public URL getURL() {
+        return sourceURL;
+    }
+
+    /**
+     * Set a string from which to read entities.
+     */
+    public void setSource(CharSequence text) {
+        source = CharSource.wrap(text);
+        sourceURL = null;
+    }
+
+    /**
+     * Set the entity format for the reader.
+     * @param format The entity format.
+     */
+    public void setFormat(EntityFormat format) {
+        this.format = format;
+    }
+
+    /**
+     * Get the entity format for the reader.
+     * @return The entity format.
+     */
+    public EntityFormat getFormat() {
+        return format;
+    }
+
+    /**
+     * Open a stream to read entities from this source.
+     * @return A stream of entities.
+     */
+    @Override
+    public ObjectStream<Entity> openStream() throws IOException {
+        BufferedReader reader = source.openBufferedStream();
+        ObjectStream<String> lines = new LineStream(reader);
+        int headerLines = format.getHeaderLines();
+        List<String> header = new ArrayList<>();
+        while (header.size() < headerLines) {
+            String line = lines.readObject();
+            if (line == null) {
+                IOException ex = new IOException(String.format("expected %d header lines, found %d", headerLines, header.size()));
+                try {
+                    lines.close();
+                } catch (Throwable th) {
+                    ex.addSuppressed(th);
```

Comment:
To my knowledge, this is exactly the code that Java would generate if we used `try`-with-resources. However, since in the non-error case we return an (unclosed) stream, we can't actually use `try`-with-resources.